File: backend/app/core/security.py
from datetime import datetime, timedelta
from typing import Optional, Dict, Any
from jose import JWTError, jwt
from passlib.context import CryptContext
from app.core.config import settings

pwd_context = CryptContext(schemes=["bcrypt"], deprecated="auto")

def verify_password(plain_password: str, hashed_password: str) -> bool:
    """Verify a password against a hash."""
    return pwd_context.verify(plain_password, hashed_password)

def get_password_hash(password: str) -> str:
    """Generate password hash."""
    return pwd_context.hash(password)

def create_access_token(data: Dict[str, Any], expires_delta: Optional[timedelta] = None) -> str:
    """Create JWT access token."""
    to_encode = data.copy()
    if expires_delta:
        expire = datetime.utcnow() + expires_delta
    else:
        expire = datetime.utcnow() + timedelta(minutes=settings.ACCESS_TOKEN_EXPIRE_MINUTES)
    
    to_encode.update({"exp": expire, "type": "access"})
    encoded_jwt = jwt.encode(to_encode, settings.SECRET_KEY, algorithm=settings.ALGORITHM)
    return encoded_jwt

def create_refresh_token(data: Dict[str, Any]) -> str:
    """Create JWT refresh token."""
    to_encode = data.copy()
    expire = datetime.utcnow() + timedelta(days=settings.REFRESH_TOKEN_EXPIRE_DAYS)
    to_encode.update({"exp": expire, "type": "refresh"})
    encoded_jwt = jwt.encode(to_encode, settings.SECRET_KEY, algorithm=settings.ALGORITHM)
    return encoded_jwt

def decode_token(token: str) -> Optional[Dict[str, Any]]:
    """Decode and verify JWT token."""
    try:
        payload = jwt.decode(token, settings.SECRET_KEY, algorithms=[settings.ALGORITHM])
        return payload
    except JWTError:
        return None

def verify_google_token(token: str) -> Optional[Dict[str, Any]]:
    """Verify Google OAuth token."""
    from google.auth.transport import requests
    from google.oauth2 import id_token
    
    try:
        print(f"🔍 Verifying Google token with CLIENT_ID: {settings.GOOGLE_CLIENT_ID[:20]}...")
        
        if not settings.GOOGLE_CLIENT_ID:
            print("❌ GOOGLE_CLIENT_ID is not set in environment variables!")
            return None
        
        idinfo = id_token.verify_oauth2_token(
            token,
            requests.Request(),
            settings.GOOGLE_CLIENT_ID
        )
        
        if idinfo['iss'] not in ['accounts.google.com', 'https://accounts.google.com']:
            print(f"❌ Invalid issuer: {idinfo['iss']}")
            return None
        
        print(f"✅ Google token verified for: {idinfo.get('email')}")
        return {
            "email": idinfo.get("email"),
            "name": idinfo.get("name"),
            "picture": idinfo.get("picture"),
            "google_id": idinfo.get("sub")
        }
    except ValueError as e:
        print(f"❌ Google token verification error: {str(e)}")
        return None
    except Exception as e:
        print(f"❌ Unexpected error verifying Google token: {str(e)}")
        return None

File: backend/src/utils/security.ts
import jwt, { SignOptions } from 'jsonwebtoken';
import bcrypt from 'bcryptjs';
import { OAuth2Client } from 'google-auth-library';
import { config } from '../config/config';

// ─── Password helpers ────────────────────────────────────────────────────────

export function verifyPassword(plain: string, hashed: string): boolean {
  return bcrypt.compareSync(plain, hashed);
}

export function getPasswordHash(password: string): string {
  return bcrypt.hashSync(password, 10);
}

// ─── JWT helpers ─────────────────────────────────────────────────────────────

export interface TokenPayload {
  sub: string;
  role?: string;
  type?: 'access' | 'refresh';
  exp?: number;
  [key: string]: unknown;
}

export function createAccessToken(data: Omit<TokenPayload, 'type' | 'exp'>): string {
  const payload: TokenPayload = { ...data, type: 'access' } as TokenPayload;
  const options: SignOptions = {
    expiresIn: config.accessTokenExpireMinutes * 60,
    algorithm: config.algorithm,
  };
  return jwt.sign(payload, config.secretKey, options);
}

export function createRefreshToken(data: Omit<TokenPayload, 'type' | 'exp'>): string {
  const payload: TokenPayload = { ...data, type: 'refresh' } as TokenPayload;
  const options: SignOptions = {
    expiresIn: config.refreshTokenExpireDays * 24 * 60 * 60,
    algorithm: config.algorithm,
  };
  return jwt.sign(payload, config.secretKey, options);
}

export function decodeToken(token: string): TokenPayload | null {
  try {
    const payload = jwt.verify(token, config.secretKey, {
      algorithms: [config.algorithm],
    }) as TokenPayload;
    return payload;
  } catch {
    return null;
  }
}

// ─── Google OAuth verification ────────────────────────────────────────────────

export interface GoogleUserInfo {
  email: string;
  name: string;
  picture?: string;
  google_id: string;
}

export async function verifyGoogleToken(token: string): Promise<GoogleUserInfo | null> {
  try {
    if (!config.googleClientId) {
      console.error('❌ GOOGLE_CLIENT_ID is not set');
      return null;
    }

    const client = new OAuth2Client(config.googleClientId);
    const ticket = await client.verifyIdToken({
      idToken: token,
      audience: config.googleClientId,
    });

    const payload = ticket.getPayload();
    if (!payload) return null;

    const iss = payload.iss || '';
    if (!['accounts.google.com', 'https://accounts.google.com'].includes(iss)) {
      console.error(`❌ Invalid issuer: ${iss}`);
      return null;
    }

    console.log(`✅ Google token verified for: ${payload.email}`);
    return {
      email: payload.email ?? '',
      name: payload.name ?? '',
      picture: payload.picture,
      google_id: payload.sub,
    };
  } catch (err) {
    console.error(`❌ Google token verification error: ${err}`);
    return null;
  }
}
